experimental/Schemes/SpecOpen.jl

```diff
@@ -0,0 +1,694 @@
+  if !issubset(Z, X) 
+    Z = intersect(X, Z)
+  end
+  return SpecOpen(X, modulus(OO(Z)))
+end
+
+SpecOpen(X::Spec) = SpecOpen(X, [one(base_ring(OO(X)))])
+
+function complement(U::SpecOpen) 
+  if !isdefined(U, :complement)
+    I = radical(saturated_ideal(ideal(localized_ring(OO(ambient(U))), gens(U))))
+    U.complement = subscheme(ambient(U), I)
+  end
+  return U.complement
+end
+
+@Markdown.doc """
+    patches(U::SpecOpen)
+
+Returns a list of principal affine open subschemes covering ``U``.
+"""
+function patches(U::SpecOpen)
+  if !isdefined(U, :patches)
+    X = ambient(U)
+    V = Vector{typeof(X)}()
+    for f in gens(U)
+      push!(V, hypersurface_complement(X, f))
+    end
+    U.patches = V
+  end
+  return U.patches
+end
+
+@Markdown.doc """
+    intersections(U::SpecOpen)
+
+Returns a list of pairwise intersections of the 
+principal open subschemes covering ``U``.
+"""
+function intersections(U::SpecOpen)
+  if !isdefined(U, :intersections)
+    X = ambient(U)
+    V = patches(U)
+    for i in 2:length(V)
+      for j in 1:i-1
+	U.intersections[(i,j)] = U.intersections[(j,i)] = intersect(V[i], V[j])
+      end
+    end
+  end
+  return U.intersections
+end
+
+function name(U::SpecOpen) 
+  if isdefined(U, :name)
+    return U.name
+  end
+  return "open subset of $(ambient(U))"
+end
+
+function intersect(
+    Y::Spec{BRT, BRET, RT, RET, MST}, 
+    U::SpecOpen{BRT, BRET, RT, RET, MST}
+  ) where {BRT, BRET, RT, RET, MST}
+  X = ambient(U)
+  if !issubset(Y, X)
+    Y = intersect(Y, X)
+  end
+  return SpecOpen(Y, gens(U))
+end
+
+function intersect(
+    U::SpecOpen{BRT, BRET, RT, RET, MST},
+    Y::Spec{BRT, BRET, RT, RET, MST}
+  ) where {BRT, BRET, RT, RET, MST}
+  return intersect(Y, U)
+end
+
+function intersect(
+    U::SpecOpen{BRT, BRET, RT, RET, MST},
+    V::SpecOpen{BRT, BRET, RT, RET, MST}
+  ) where {BRT, BRET, RT, RET, MST}
+  X = ambient(U) 
+  X == ambient(V) || error("method not implemented")
+  return SpecOpen(X, [a*b for a in gens(U) for b in gens(V)])
+end
+
+function Base.union(U::T, V::T) where {T<:SpecOpen}
+  ambient(U) == ambient(V) || error("the two open sets do not lay in the same ambient variety")
+  return SpecOpen(ambient(U), vcat(gens(U), gens(V)))
+end
+
+function issubset(
+    Y::Spec{BRT, BRET, RT, RET, MST}, 
+    U::SpecOpen{BRT, BRET, RT, RET, MST}
+  ) where {BRT, BRET, RT, RET, MST}
+  return one(localized_ring(OO(Y))) in ideal(OO(Y), gens(U))
+end
+
+function issubset(
+    U::SpecOpen{BRT, BRET, RT, RET, MST},
+    Y::Spec{BRT, BRET, RT, RET, MST}
+  ) where {BRT, BRET, RT, RET, MST}
+  for V in patches(U)
+    issubset(V, Y) || return false
+  end
+  return true
+end
+
+function issubset(U::T, V::T) where {T<:SpecOpen}
+  base_ring(OO(ambient(U))) == base_ring(OO(ambient(V))) || return false
+  return issubset(complement(intersect(V, ambient(U))), complement(U))
+end
+
+function ==(U::T, V::T) where {T<:SpecOpen}
+  return issubset(U, V) && issubset(V, U)
```

Comment:
I don't get that.

Reply:
I suppose that `issubset` can be expensive. But if `U === V` then checking twice `issubset`  is not necessary,

Reply:
@simonbrandhorst Perhaps the `issubset` methods should have this optimization instead?

@HechtiDerLachs what do you mean?

Reply:
I didn't get what the code is supposed to do and why that would be useful. Is `!==` the negation of `===` which compares the addresses? Then, this would be a shortcut, indeed.

Reply:
Indeed `!==` relates to `===` just as `!=` relates to `==` (this can also be checked via the built-in REPL help)